Simplify slot allocation

henriknyman opened the pull request:
Makes allocating a slot for a variable an idempotent operation,
so that individual steps in the slot allocation pass does not need
to worry about if a slot has already been allocated for a variable
within the same pipepline by a previous step.

The slot may have slightly different nullability and specific type as
long as it is compatible.
- If nullability differs, the nullability is conservatively increased.
- If types differs (but are compatible) the least specific type
is chosen.

If the slots are incompatible a compile-time exception is still thrown.

This also fixes CartesianProduct, ValueHashJoin and NodeHashJoin to work
with shared argument slots, and a bug in slot allocation for arguments.

henriknyman committed:
Simplify slot allocation

Makes allocating a slot for a variable an idempotent operation,
so that individual steps in the slot allocation pass does not need
to worry about if a slot has already been allocated for a variable
within the same pipepline by a previous step.

The slot may have slightly different nullability and specific type as
long as it is compatible.
- If nullability differs, the nullability is conservatively increased.
- If types differs (but are compatible) the least specific type
is chosen.

If the slots are incompatible a compile-time exception is still thrown.

fickludd commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotConfiguration.scala:
```diff
@@ -189,21 +188,73 @@ class SlotConfiguration(private val slots: mutable.Map[String, Slot],
+      slots.put(key, slot)
+      slotAliases.addBinding(slot, key)
+      numberOfLongs = numberOfLongs + 1
+    }
```
An alternative way on handling this option would be

```
slots.get(key) match {
  case Some(existingSlot) =>
      ...
  case None =>
      ...
}
```

fickludd commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotConfiguration.scala:
```diff
@@ -189,21 +188,73 @@ class SlotConfiguration(private val slots: mutable.Map[String, Slot],
+      slots.put(key, slot)
+      slotAliases.addBinding(slot, key)
+      numberOfReferences = numberOfReferences + 1
+    }
```
See `Option[Slot]` match above.

fickludd commented on enterprise/cypher/physical-planning/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotConfigurationTest.scala:
```diff
@@ -0,0 +1,152 @@
+    // then
+    slots("a") should equal(RefSlot(0, nullable = false, CTNode))
+    clone.get("a") shouldBe empty
+    clone.numberOfReferences should equal(0)
```
I really like this test. It would be even better though if it had some more assertions of that the longs are preserved, and that aliases are still the same.

henriknyman replied:
I just moved this test, but sure, I'll add some more assertions

fickludd replied:
nice!

fickludd reviewed (changes requested):
I really like this change in `SlotConfiguration`. And especially the type inference. Also really like the test coverage. Just requesting some more assertions on the deepClone test, then I'm happy!

henriknyman committed:
Fix some review comments

- Changed if to match
- Added more test assertions

fickludd committed:
Resolve to test issues.

systay commented:
Hmm... I must be missing something here. When would two different operators in the same pipeline want to declare the same variable? I don't understand the context of when this would be used.

fickludd commented:
@systay this is not common but it happened in a few places in the spike if I remember correctly. On merge relationship it happens every time because of the explicit branching and locking.

Today I discovered that this PR does not play nice with the `ValueHashJoin` so we'll have to think about that some more before merging this.

henriknyman committed:
Fix HashJoins and CartesianProduct to handle overlapping argument slots

Adds an assertion on the assumption that the only shared slots
in NodeHashJoin, ValueHashJoin and CartesianProduct are argument slots
that are identical on both sides. In the pipes we then exclude these
arguments when copying slots from the right hand side.

henriknyman committed:
Fix bug in slot allocation of arguments for leaf plans

When allocating slots for leaf plans the slot configuration from
the argument needs to be copied, or else we end up incorrectly
adding slots from leaf plans to the slot configuration of both sides
of branched plans like CartesianProduct or HashJoins.

henriknyman committed:
Update test that is now succeeding
